subseas_flex_fcst/cpt.py

```diff
@@ -0,0 +1,95 @@
+    Returns
+    -------
+    file_selected : xarray Dataset
+        Single CPT data file as multidimensional xarray dataset.
+    Notes
+    -----
+    `filename_pattern` should be most common denominator for any group of datasets,
+    so that a single file can be selected using only `lead_time` and `start_date`.
+    Examples
+    --------
+    For files which have naming structure such as the example file:
+        CFSv2_SubXPRCP_CCAFCST_mu_Apr_Apr-1-2022_wk1.txt
+    And where this file's `lead_time` and `start_date`:
+        `lead_time` == 'wk1' and `start_date` == 'Apr-1-2022'
+    `filename_pattern` == 'CFSv2_SubXPRCP_CCAFCST_mu_Apr_*.txt'
+    """
+    pattern = f"{start_date}_{lead_time}"
+    full_path = f"{data_path}/{filename_pattern}"
+    expanded_name = glob.glob(full_path.replace("mystartandlead",pattern))
+    if len(expanded_name) == 0:
+        file_selected = xr.Dataset().merge(xr.DataArray(np.nan).rename("missing_file"))
+    else:
+        file_name = expanded_name[0]
+        file_selected = cptio.open_cptdataset(file_name)
+    return file_selected
+
+
+def starts_list(
+    data_path,
+    filename_pattern,
+    regex_search_pattern,
+    format_in="%b-%d-%Y",
+    format_out="%b-%-d-%Y",
+):
+    """ Get list of all start dates from CPT files.
+
+    Parameters
+    ----------
+    data_path : str
+        String of the path pointing to cpt datasets.
+    filename_pattern : str
+        String of the filename pattern name for a given variable's data file.
+    regex_search_pattern : str
+        String representing regular expression search pattern to find dates in file names.
+    format_in : str
+        String representing dates format found in file names
+    format_out : str
+        String representing desired output dates format.
+    Returns
+    -------
+    start_dates : list
+        List of strings representing all start dates for the data within `data_path`.
+    Notes
+    -----
+    For more information on regex visit: https://docs.python.org/3/library/re.html
+    Test your regex code here: https://regexr.com/
+    Examples
+    --------
+    Regex expression "\w{3}-\w{1,2}-\w{4}" matches expressions that are:
+    '{word of 3 chars}-{word between 1,2 chars}-{word of 4 chars}'
+    will match dates of format 'Apr-4-2022', 'dec-14-2022', etc.
+    """
+    filename_pattern = filename_pattern.replace("mystartandlead", "*")
+    files_name_list = glob.glob(f'{data_path}/{filename_pattern}')
+    start_dates = []
+    for file in files_name_list:
+        start_date = re.search(regex_search_pattern, file)
+        start_date_dt = datetime.strptime(start_date.group(), format_in)
+        start_dates.append(start_date_dt)
+    start_dates = sorted(set(start_dates)) #finds unique dates in the case there are files with the same date due to multiple lead times
+    start_dates = [i.strftime(format_out) for i in start_dates]
+    return start_dates
```

Comment:
Seeing this CPT-specific stuff makes me lean more towards zarrification than I did when you asked earlier today. I'm ok with this version, but if you wanted to convert to zarr first and then use normal xarray operations instead of `sel_file` and `starts_list`, I would be ok with that too.

Reply:
That's a bit beyond what I can afford to do with this PR now.

Reply:
yeah... this will have to be considered at another time/PR I am afraid.